v-next/hardhat-viem/src/internal/clients.ts

```diff
@@ -0,0 +1,162 @@
+import type {
+  GetPublicClientReturnType,
+  GetWalletClientReturnType,
+  TestClient,
+} from "../types.js";
+import type { ChainType } from "@ignored/hardhat-vnext/types/network";
+import type { EthereumProvider } from "@ignored/hardhat-vnext/types/providers";
+import type {
+  Address as ViemAddress,
+  PublicClientConfig as ViemPublicClientConfig,
+  TestClientConfig as ViemTestClientConfig,
+  WalletClientConfig as ViemWalletClientConfig,
+} from "viem";
+
+import { HardhatError } from "@ignored/hardhat-vnext-errors";
+import {
+  createPublicClient,
+  createWalletClient,
+  createTestClient,
+  custom as customTransport,
+} from "viem";
+import { publicActionsL2, walletActionsL2 } from "viem/op-stack";
+
+import { getAccounts } from "./accounts.js";
+import { getChain, getMode, isDevelopmentNetwork } from "./chains.js";
+
+export async function getPublicClient<ChainTypeT extends ChainType | string>(
+  provider: EthereumProvider,
+  chainType: ChainTypeT,
+  publicClientConfig?: Partial<ViemPublicClientConfig>,
+): Promise<GetPublicClientReturnType<ChainTypeT>> {
+  const chain = publicClientConfig?.chain ?? (await getChain(provider));
+  const parameters = {
+    ...getDefaultClientParameters(chain.id),
+    ...publicClientConfig,
+  };
+
+  let publicClient = createPublicClient({
+    chain,
+    transport: customTransport(provider),
+    ...parameters,
+  });
+
+  if (chainType === "optimism") {
+    publicClient = publicClient.extend(publicActionsL2());
+  }
+
+  /* eslint-disable-next-line @typescript-eslint/consistent-type-assertions --
+  We need to assert the type because TS gets confused with the conditional type */
+  return publicClient as GetPublicClientReturnType<ChainTypeT>;
```

Comment:
I took a brief look here, but have no suggestions.
Do we consider this technical debt or just a limitation in TS? @alcuadrado

Reply:
It turns our there are limitations in typescript.